zerver/lib/mention.py

```diff
@@ -37,27 +37,58 @@ def Q(self) -> Q:
             raise AssertionError("totally empty filter makes no sense")
 
 
-@dataclass
 class MentionBackend:
-    realm_id: int
+    def __init__(self, realm_id: int) -> None:
+        self.realm_id = realm_id
+        self.cache: Dict[Tuple[int, str], FullNameInfo] = {}
 
     def get_full_name_info_list(self, user_filters: List[UserFilter]) -> List[FullNameInfo]:
-        q_list = [user_filter.Q() for user_filter in user_filters]
+        result: List[FullNameInfo] = []
+        new_user_filters: List[UserFilter] = []
 
-        rows = (
-            UserProfile.objects.filter(
-                realm_id=self.realm_id,
-                is_active=True,
-            )
-            .filter(
-                functools.reduce(lambda a, b: a | b, q_list),
-            )
-            .only(
-                "id",
-                "full_name",
+        # Try to get messages from the cache first.
+        for user_filter in user_filters:
+            user: Optional[FullNameInfo] = None
+
+            if self.cache:
+                # We expect callers who take advantage of our cache to supply both
+                # id and full_name in the user mentions in their messages.
+                if user_filter.id is not None and user_filter.full_name is not None:
+                    user = self.cache.get((user_filter.id, user_filter.full_name), None)
+
+            if user is not None:
+                result.append(user)
```

Comment:
Could this be written more readably with `if self.cache:` and then doing e.g. `assert user_filter.id is not None`, moving this if statement inside the `self.cache` block, and doing a `continue` after the `result.append` line rather than an else clause?  I was having trouble thinking about all the fallthrough situations in this function.

Reply:
Hmm I guess the assert part doesn't work since callers aren't explicit about whether they want the cache, but the rest of the ideas seem helpful.

Reply:
I don't understand the `continue` suggestion.  The idea in the first loop is that you either get a cache hit (and append to results) or a cache miss (and then need to add to the list of things to do in the second loop).  I tried tweaking a few variables and comments here, so hopefully on the next review the code will make a bit more sense.

Reply:
I believe the suggestion is

```diff
 for user_filter in user_filters:
-    user: Optional[FullNameInfo] = None
-
     if user_filter.id is not None and user_filter.full_name is not None:
         user = self.user_cache.get((user_filter.id, user_filter.full_name), None)
+        if user is not None:
+            result.append(user)
+            continue
 
-    if user is not None:
-        result.append(user)
-    else:
-        unseen_user_filters.append(user_filter)
+    unseen_user_filters.append(user_filter)
```

Reply:
Ah, ok, fix coming.

Reply:
re-pushed